core/chains/evm/txmgr/nonce_tracker.go

```diff
@@ -0,0 +1,201 @@
+	if err != nil || len(txs) == 0 {
+		return seq, fmt.Errorf("failed to get transactions from tx store for address (%s): %w", address.String(), err)
+	}
+
+	foundSeq := false
+	for _, tx := range txs {
+		if tx.Sequence == nil {
+			continue
+		}
+		foundSeq = true
+		if *tx.Sequence > seq {
+			seq = *tx.Sequence
+		}
+	}
+
+	// Unstarted transactions can be part of the list which would not have nonces assigned
+	// Only want to return successfully if a sequence is found in the tx store, otherwise return error to check on-chain
+	if !foundSeq {
+		return seq, fmt.Errorf("failed to extract nonce from transactions for address: %s", address.String())
+	}
+
+	seq++ // Need to increment since the current one is already in use
+	return seq, nil
+}
+
+// syncSequence tries to sync the key sequence, retrying indefinitely until success or stop signal is sent
+func (s *nonceTracker) SyncSequence(ctx context.Context, addr common.Address, chStop services.StopChan) {
+	sequenceSyncRetryBackoff := backoff.Backoff{
+		Min:    100 * time.Millisecond,
+		Max:    5 * time.Second,
+		Jitter: true,
+	}
+
+	localSequence, err := s.GetNextSequence(ctx, addr)
+	// Address not found in map so skip sync
+	if err != nil {
+		s.lggr.Criticalw("Failed to retrieve local next sequence for address", "address", addr.String(), "err", err)
+		return
+	}
+
+	// Enter loop with retries
+	var attempt int
+	for {
+		select {
+		case <-chStop:
+			return
+		case <-time.After(sequenceSyncRetryBackoff.Duration()):
+			attempt++
+			err := s.SyncOnChain(ctx, addr, localSequence)
+			if err != nil {
+				if attempt > 5 {
+					s.lggr.Criticalw("Failed to sync with on-chain sequence", "address", addr.String(), "attempt", attempt, "err", err)
+				} else {
+					s.lggr.Warnw("Failed to sync with on-chain sequence", "address", addr.String(), "attempt", attempt, "err", err)
+				}
+				continue
+			}
+			return
+		}
+	}
+}
+
+func (s *nonceTracker) SyncOnChain(ctx context.Context, addr common.Address, localSequence evmtypes.Nonce) error {
+	nonce, err := s.client.PendingSequenceAt(ctx, addr)
+	if err != nil {
+		return err
+	}
+	if nonce > localSequence {
+		s.lggr.Warnw(fmt.Sprintf("address %s has been used before, either by an external wallet or a different Chainlink node. "+
+			"Local nonce is %v but the on-chain nonce for this account was %v. "+
+			"It's possible that this node was restored from a backup. If so, transactions sent by the previous node will NOT be re-org protected and in rare cases may need to be manually bumped/resubmitted. "+
+			"Please note that using the chainlink keys with an external wallet is NOT SUPPORTED and can lead to missed or stuck transactions. ",
+			addr, localSequence, nonce),
+			"address", addr.String(), "localNonce", localSequence, "chainNonce", nonce)
+
+		s.lggr.Infow("Fast-forward sequence", "address", addr, "newNextSequence", nonce, "oldNextSequence", localSequence)
+	}
+
+	s.sequenceLock.Lock()
```

Comment:
It's a bit dangerous to lock after doing a check that can affect whether we want to proceed with the write. In this case, what happens if the value of `nonce` here changes after you do the check on 152 but before you lock here?

Reply:
The check on line 152 is actually only used to determine if we should log some statements, otherwise we proceed with the write regardless. Also the `nonce` here is just the one we retrieved from on chain. I guess it could change but the sequence map lock wouldn't prevent any desync with on-chain data.

Reply:
Ok cool this makes sense. So it's fine as is then!